src/Yesod/Page/RenderedRoute.hs

```diff
@@ -0,0 +1,99 @@
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE RecordWildCards #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+
+module Yesod.Page.RenderedRoute
+  ( RenderedRoute
+  , getRenderedRoute
+  , updateQueryParameter
+  )
+where
+
+import Data.Aeson
+import Data.Map (Map)
+import qualified Data.Map as Map
+import Data.Maybe (fromMaybe)
+import Data.Text (Text, intercalate)
+import UnliftIO (throwString)
+import Yesod.Core
+  ( HandlerSite
+  , MonadHandler
+  , RenderRoute
+  , getCurrentRoute
+  , getRequest
+  , renderRoute
+  , reqGetParams
+  )
+
+-- | Information about the initial route
+data RenderedRoute = RenderedRoute
+  { renderedRouteDomain :: Maybe Text
+  , renderedRoutePath :: [Text]
+  , renderedRouteQuery :: [(Text, Text)]
+  }
+
+instance ToJSON RenderedRoute where
+  -- | Renders as a @'Text'@ URL
+  --
+  -- FIXME: This is wildly naive and we should move to a proper library to
+  -- ensure escaping is handled, but it's meant only to prove the approach and
+  -- pass the current test suite.
+  --
+  toJSON RenderedRoute {..} = String
+    $ fromMaybe "" renderedRouteDomain
+    <> intercalate "/" renderedRoutePath
+    <> renderQuery renderedRouteQuery
+   where
+    renderQuery [] = ""
+    renderQuery qs = "?" <> intercalate "&" (map (\(k, v) -> k <> "=" <> v) qs)
+
+-- | Get the current route as a @'RenderedRoute'@
+--
+-- TODO: supplying the domain is a transitional thing to ease the diff with
+-- previous code, we can just grab Host off the @'Request'@ here, I think.
+--
+-- That would:
+--
+-- 1. Simplify @'PageConfig'@ further
+-- 2. Be more robust to to different Approot configurations
+-- 3. Be more directly our intentions (IMO), which are: visit this same Host
+--    but with these updated parameters
+--
+getRenderedRoute
+  :: (MonadHandler m, RenderRoute (HandlerSite m))
+  => Maybe Text
+  -> m RenderedRoute
+getRenderedRoute domain = do
+  route <- maybe (throwString "no route") pure =<< getCurrentRoute
+
+  -- When I just use _query, it's always empty. Why would renderRoute return
+  -- this tuple if the Route value (apparently) never has the query information?
+  let (path, _query) = renderRoute route
+  query <- reqGetParams <$> getRequest
+
+  pure $ RenderedRoute
+    { renderedRouteDomain = domain
+    , renderedRoutePath = path
+    , renderedRouteQuery = query
+    }
+
+-- | Update a single query parameter and preserve the rest
+--
+-- If given @'Nothing'@, the parameter is removed.
+--
+updateQueryParameter :: Text -> Maybe Text -> RenderedRoute -> RenderedRoute
+updateQueryParameter name = overQuery . asMap . updateKey name
```

Comment:
If you wanted less intermediate values here then you could `map` over the `[(Text, Text)]` and only replace a value when matching the given param.

```
map f params

f find replace x@(label, _)
  | label == find = (label, replace)
  | otherwise = x
```

Reply:
Oh yeah, that's nice.

Reply:
This loses the feature of removing the param if given `Nothing`, or am I misunderstanding?

Reply:
So, I wrote a version of your code that does support the `Nothing` deletion, and it fails the tests. I can't see why:

```hs
updateKey :: Eq a => a -> Maybe b -> [(a, b)] -> [(a, b)]
updateKey find mReplace = mapMaybe setKey
 where
  setKey x@(label, _)
    | label == find = (label, ) <$> mReplace
    | otherwise = Just x
```

I then wrote this code and it passes the tests:

```hs
updateKey :: (Ord a, Ord b) => a -> Maybe b -> [(a, b)] -> [(a, b)]
updateKey k mv = maybe id (\v -> List.insert (k, v)) mv . deleteByKey k

deleteByKey :: Eq a => a -> [(a, b)] -> [(a, b)]
deleteByKey _ [] = []
deleteByKey k (x@(k', _) : xs)
  | k == k' = deleteByKey k xs
  | otherwise = x : deleteByKey k xs
```

I like the idea of removing the `Map` stuff, but the only non-`Map` code I could get to pass the tests is reeaaally just approximating a Map, poorly.

Can you spot the bug in the first code? I'm feeling a bit crazy...

Reply:
Yeah, that first code looks fine to me :eyes:

Reply:
Found it, it can't add a param.

Reply:
Aha! Thank you. Will keep current for now.

Reply:
So `updateQueryParameter` is really `upsertQueryParameter`.